openmls/src/group/core_group/mod.rs

```diff
@@ -751,6 +751,47 @@ impl CoreGroup {
         }
     }
 
+    pub(crate) fn own_leaf_node(&self) -> Result<&OpenMlsLeafNode, LibraryError> {
+        self.tree
+            .leaf(self.own_leaf_index())
+            .ok_or_else(|| LibraryError::custom("Tree has no own leaf."))
+    }
+}
+
+#[cfg(any(feature = "test-utils", test))]
+impl CoreGroup {
+    pub(crate) fn context_mut(&mut self) -> &mut GroupContext {
+        &mut self.group_context
+    }
+
+    pub(crate) fn message_secrets_test_mut(&mut self) -> &mut MessageSecrets {
+        self.message_secrets_store.message_secrets_mut()
+    }
+
+    pub(crate) fn print_tree(&self, message: &str) {
+        use super::tests::tree_printing::print_tree;
+
+        print_tree(self, message);
+    }
+}
+
+/// Composite key for key material of a client within an epoch
+pub struct EpochKeypairId(Vec<u8>);
+
+impl EpochKeypairId {
+    fn new(group_id: &GroupId, epoch: u64, leaf_index: LeafNodeIndex) -> Self {
+        Self(
+            [
+                group_id.as_slice(),
+                &leaf_index.u32().to_be_bytes(),
+                &epoch.to_be_bytes(),
+            ]
+            .concat(),
+        )
+    }
+}
+
+impl CoreGroup {
```

Comment:
I'm a bit confused as to the structure here. Is there a reason you interrupt the CoreGroup impl at this point? Just to introduce the `EpochKeypairId`? Also, the test-utils block was further below. Is there a reason you moved it up here?

Reply:
Reordered